### PR TITLE
Fix/215/sklearn mask nan

### LINKED ISSUE #215: Traditional ML workflow should mask `NaN` effectively
As @cynthiaxu04 observed, currently the LGBM and similar scikit-learn driven workflows jam when `NaN` values are in the targets. We should make sure this works automatically like with chemprop workflows

### PR DESCRIPTION
**Pull request by cynthiaxu04**
## Description
Closes #215. 

## Todos
  - [x] `DataSpec.read()` updated to drop `NaN`s for both Traditional ML and DL anvil workflows
  - [x] `DataSpec.read()` also now returns the original number of rows and the number of rows dropped if any. This will appear in a logger.info message

## Questions
- [ ] Question1

## Status
- [ ] Ready to go

### CODE REVIEW

File: openadmet/models/anvil/data_spec.py [hunk 1]
```diff
@@ -68,7 +68,25 @@ def read(self) -> tuple[pd.Series, pd.Series]:
...
+            n_dropped = n_before - n_after
+        else:
+            n_dropped = 0
+            cleaned_combined = combined
+
+        # Split the data again
+        input_clean = cleaned_combined[self.input_col]
+        targets_clean = cleaned_combined[self.target_cols]
+
+        return input_clean, targets_clean, n_dropped, n_before
```
Comment (smcolby) [inline openadmet/models/anvil/data_spec.py:89]: Why are we returning so much here? My expectation would be to get back X (inputs) and y (targets). Moreover, I think we generally want to broker in numpy arrays when passing data around, not pandas Series.

@hmacdope might have opinions here.

File: openadmet/models/anvil/data_spec.py [hunk 2]
```diff
@@ -68,7 +68,25 @@ def read(self) -> tuple[pd.Series, pd.Series]:
         targets = data[self.target_cols]
         input = data[self.input_col]
 
-        return input, targets
+        # combine input and targets for joint NaN handling
+        combined = pd.concat([input, targets], axis=1)
+
```
Comment (smcolby) [inline openadmet/models/anvil/data_spec.py:73]: Could be replaced by:
```
combined = data[self.input_cols + self.target_cols]
```
You're otherwise subsetting the data and recombining.

File: openadmet/models/anvil/workflow.py [hunk 3]
```diff
@@ -381,7 +381,8 @@ def run(
 
         # Load data from YAML specification
         logger.info("Loading data")
-        X, y = self.data_spec.read()
+        X, y, n_dropped, n_before = self.data_spec.read()
+        logger.info(f"{n_before} total rows. {n_dropped} NaN rows were dropped.")
```
Comment (smcolby) [inline openadmet/models/anvil/workflow.py:385]: This might answer my question: `n_dropped` and `n_before` are used for logging, but this info is only accessible if it "comes out of" `data_spec.read()`. 

Feels a bit hacky to save this information out just to print a log message, particularly when `dropna` isn't necessarily going to be set to `True`. Might I suggest we add dropping NaNs as a data filter and separate step, rather than trying to force it into the load functionality? Or simply log "Dropping NaN values" without a count in the output?

File: openadmet/models/anvil/workflow.py [hunk 4]
```diff
@@ -381,7 +381,8 @@ def run(
 
         # Load data from YAML specification
         logger.info("Loading data")
-        X, y = self.data_spec.read()
+        X, y, n_dropped, n_before = self.data_spec.read()
```
Comment (smcolby) [inline openadmet/models/anvil/workflow.py:384]: `data_spec.read()` is not passed a `dropna` argument, so NaNs won't ever be dropped here.

File: openadmet/models/tests/data/test_data.py [hunk 5]
```diff
@@ -23,6 +23,6 @@ def test_data_spec_from_intake():
         target_cols=["data1"],
         input_col="SMILES",
     )
-    target, smiles = data_spec.read()
+    target, smiles, _, _ = data_spec.read()
     assert len(target) == 30
     assert len(smiles) == 30
```
Comment (smcolby) [inline openadmet/models/tests/data/test_data.py:28]: Should have tests for when `dropna` is `True`

**Review by smcolby — CHANGES_REQUESTED** [review]

File: openadmet/models/anvil/data_spec.py (hunk 1 above)
Comment (hmacdope) [reply openadmet/models/anvil/data_spec.py:89]: See below comment on logging

File: openadmet/models/anvil/workflow.py (hunk 4 above)
Comment (cynthiaxu04) [reply openadmet/models/anvil/workflow.py:384]: `dropna` is defaulted to True

File: openadmet/models/anvil/data_spec.py [hunk 6]
```diff
@@ -64,11 +66,28 @@ def read(self) -> tuple[pd.Series, pd.Series]:
...
         else:
             raise ValueError(f"Unsupported resource type: {self.resource}")
-        # now read the target columns and smiles column
-        targets = data[self.target_cols]
-        input = data[self.input_col]
 
-        return input, targets
+        # combine input and targets for joint NaN handling
+        combined = data[[self.input_col] + (self.target_cols if isinstance(self.target_cols, list) else [self.target_cols])]
+
```
Comment (smcolby) [inline openadmet/models/anvil/data_spec.py:72]: Nice.

File: openadmet/models/anvil/data_spec.py [hunk 7]
```diff
@@ -64,11 +66,28 @@ def read(self) -> tuple[pd.Series, pd.Series]:
...
+        # get number of combined rows for logging
+        n_before = len(combined)
+
+        if self.dropna:
+            cleaned_combined=combined.dropna()
```
Comment (smcolby) [inline openadmet/models/anvil/data_spec.py:77]:
```suggestion
            cleaned_combined = combined.dropna()
```

File: openadmet/models/anvil/data_spec.py [hunk 8]
```diff
@@ -64,11 +66,28 @@ def read(self) -> tuple[pd.Series, pd.Series]:
...
+            n_dropped = 0
+            cleaned_combined = combined
+
+        # Split the data again
+        input_clean = cleaned_combined[self.input_col]
+        targets_clean = cleaned_combined[self.target_cols]
+
+        logger.info(f"{n_before} total rows. {n_dropped} NaN rows were dropped.")
+
+        return input_clean, targets_clean
```
Comment (smcolby) [inline openadmet/models/anvil/data_spec.py:90]: Do we want these returned as data frames / series, or as numpy arrays? I.e. `input_clean.values, targets_clean.values`.

File: openadmet/models/tests/data/test_data.py [hunk 9]
```diff
@@ -26,3 +26,26 @@ def test_data_spec_from_intake():
...
+        input_col="OPENADMET_CANONICAL_SMILES",
+        dropna=False
+    )
+
+    target, smiles = data_spec.read()
+    target2, smiles2, = data_spec2.read()
+    assert len(target) == 3333
+    assert len(smiles) == 3333
+    assert len(target2) == 7196
+    assert len(smiles2) == 7196
```
Comment (smcolby) [inline openadmet/models/tests/data/test_data.py:51]: This will test both of your cases independently with fewer lines of code.

```python
@pytest.mark.parametrize("dropna, expected_length", [(True, 3333), (False, 7196)])
def test_data_spec_dropna(dropna, expected_length):
    data_spec = DataSpec(
        type="intake",
        resource=nan_data,
        target_cols=["OPENADMET_LOGAC50"],
        input_col="OPENADMET_CANONICAL_SMILES",
        dropna=dropna
    )

    target, smiles = data_spec.read()

    assert len(target) == expected_length
    assert len(smiles) == expected_length
```

**Review by smcolby — CHANGES_REQUESTED** [review]
Almost there!

File: openadmet/models/anvil/data_spec.py (hunk 8 above)
Comment (cynthiaxu04) [reply openadmet/models/anvil/data_spec.py:90]: the read function has the return defined as a tuple of pandas series

Comment (smcolby) [reply openadmet/models/anvil/data_spec.py:90]: I'm referring to #130 here

**Review by smcolby — APPROVED** [review]
All good!

File: openadmet/models/anvil/data_spec.py (hunk 8 above)
Comment (cynthiaxu04) [reply openadmet/models/anvil/data_spec.py:90]: has to do with issue #130
